src/main/drivers/bus_spi_hal.c

```diff
@@ -91,128 +128,122 @@ void spiInitDevice(SPIDevice device, bool leadingEdge)
         // Rx only
-        status = HAL_SPI_Receive(&spiDevice[device].hspi, in, len, SPI_TIMEOUT_SYS_TICKS);
+        status = HAL_SPI_Receive(&spiDevice[device].hspi, rxData, len, SPI_TIMEOUT_SYS_TICKS);
     } else {
         // Tx and Rx
-        status = HAL_SPI_TransmitReceive(&spiDevice[device].hspi, out, in, len, SPI_TIMEOUT_SYS_TICKS);
+        status = HAL_SPI_TransmitReceive(&spiDevice[device].hspi, txData, rxData, len, SPI_TIMEOUT_SYS_TICKS);
     }
 
-    if (status != HAL_OK) {
-        spiTimeoutUserCallback(instance);
-    }
-
-    return true;
+    return (status == HAL_OK);
 }
 
-// Position of Prescaler bits are different from MCU to MCU
-
-static uint32_t baudRatePrescaler[8] = {
-    SPI_BAUDRATEPRESCALER_2,
-    SPI_BAUDRATEPRESCALER_4,
-    SPI_BAUDRATEPRESCALER_8,
-    SPI_BAUDRATEPRESCALER_16,
-    SPI_BAUDRATEPRESCALER_32,
-    SPI_BAUDRATEPRESCALER_64,
-    SPI_BAUDRATEPRESCALER_128,
-    SPI_BAUDRATEPRESCALER_256,
-};
-
-void spiSetDivisor(SPI_TypeDef *instance, uint16_t divisor)
+void spiInternalInitStream(const extDevice_t *dev, bool preInit)
 {
-    SPIDevice device = spiDeviceByInstance(instance);
-
-    HAL_SPI_DeInit(&spiDevice[device].hspi);
-
-    spiDevice_t *spi = &(spiDevice[device]);
-
-    int prescalerIndex = ffs(divisor) - 2; // prescaler begins at "/2"
-
-    if (prescalerIndex < 0 || prescalerIndex >= (int)ARRAYLEN(baudRatePrescaler)) {
-        return;
-    }
-
-    spi->hspi.Init.BaudRatePrescaler = baudRatePrescaler[prescalerIndex];
-
-    HAL_SPI_Init(&spi->hspi);
+    UNUSED(dev);
+    UNUSED(preInit);
 }
 
-#ifdef USE_DMA
-DMA_HandleTypeDef* dmaHandleByInstance(SPI_TypeDef *instance)
+void spiInternalStartDMA(const extDevice_t *dev)
 {
-    return &spiDevice[spiDeviceByInstance(instance)].hdma;
+    UNUSED(dev);
 }
 
-void SPI1_IRQHandler(void)
+void spiInternalStopDMA (const extDevice_t *dev)
 {
-    HAL_SPI_IRQHandler(&spiDevice[SPIDEV_1].hspi);
+    UNUSED(dev);
 }
 
-void SPI2_IRQHandler(void)
+void spiInternalResetDescriptors(busDevice_t *bus)
 {
-    HAL_SPI_IRQHandler(&spiDevice[SPIDEV_2].hspi);
+    UNUSED(bus);
 }
 
-void SPI3_IRQHandler(void)
+void spiInternalResetStream(dmaChannelDescriptor_t *descriptor)
 {
-    HAL_SPI_IRQHandler(&spiDevice[SPIDEV_3].hspi);
+    UNUSED(descriptor);
 }
 
-void SPI4_IRQHandler(void)
+// Transfer setup and start
+void spiSequence(const extDevice_t *dev, busSegment_t *segments)
 {
-    HAL_SPI_IRQHandler(&spiDevice[SPIDEV_4].hspi);
-}
+    busDevice_t *bus = dev->bus;
+    SPIDevice device = spiDeviceByInstance(bus->busType_u.spi.instance);
+    SPI_HandleTypeDef *hspi = &spiDevice[device].hspi;
+
+    bus->initSegment = true;
+    bus->curSegment = segments;
+
+    // Switch bus speed
+    spiSetDivisor(bus->busType_u.spi.instance, dev->busType_u.spi.speed);
+
+    // Switch SPI clock polarity/phase if necessary
+    if (dev->busType_u.spi.leadingEdge != bus->busType_u.spi.leadingEdge) {
+        if (dev->busType_u.spi.leadingEdge){
```

Comment:
`leadingEdge` should probably be `clockConfiguration` as uint8 flags.  so that polarity and edge can be set and this avoid the boolean argument passing, storing and checking of `leadingEdge`.